### PR TITLE
Improve Text Extraction

### PR DESCRIPTION
**Pull request by pubpub-zz**
fix at least #880

line feeds are applied as stated in the ref

TODO : for segmented text, horizontal translation should be analysed to apply some space or delete some characters (to be analysed in crazyones.pdf)
TODO : font conversion (to be analysed in crazyones.pdf)

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma 
 For review only, still some works before merging into 2.0

**Comment by pubpub-zz** [comment]
deep changes in the text extractions. still some analysis with some fonts(truetype,type 3)

Looking for so beta testers.

**Comment by MartinThoma** [comment]
Uuuh, nice! I'll let it run with https://github.com/py-pdf/benchmarks this evening if I have some time :-)

File: PyPDF2/_page.py [hunk 1]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+                    encoding = list(charset_encoding['/StandardCoding'])
+                if '/Differences' in enc:
+                    x=0
+                    for o in enc['/Differences']:
+                        if isinstance(o,int):
+                            x = o
+                        else:
+                            try:
+                                encoding[x] = adobe_glyphs[o]
+                            except:
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1088]:
```suggestion
                            except Exception:
```

File: PyPDF2/generic.py [hunk 2]
```diff
@@ -1702,7 +1708,17 @@ def createStringObject(
 
     :raises TypeError: If string is not of type str or bytes.
     """
-    if isinstance(string, str):
+    if isinstance(forceEncoding,list):
+        out = ""
+        for x in string:
+            try:
+                out += forceEncoding[x]
+            except:
```
Comment (MartinThoma) [inline PyPDF2/generic.py:1716]:
```suggestion
            except Exception:
```

File: PyPDF2/generic.py [hunk 3]
```diff
@@ -2015,14 +2031,68 @@ def decode_pdfdocencoding(byte_array: bytes) -> str:
...
-        str(char) + " at " + str(i) + " already at " + str(_pdfDocEncoding_rev[char])
-    )
-    _pdfDocEncoding_rev[char] = i
+assert len(_pdfdoc_encoding) == 256
+def fill_from_encoding(enc:str)->list:
+    lst=()
+    for x in range(256):
+        try:
+            lst+=(bytes((x,)).decode(enc),)
+        except:
```
Comment (MartinThoma) [inline PyPDF2/generic.py:2040]:
```suggestion
        except Exception:
```

File: PyPDF2/_page.py [hunk 4]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
 
         :return: a string object.
         """
+        # code freely inspired from @twiggy ; see #711
+        def buildCharMap(pdf, font_name="/a"):
+            mapDict = {}
+            processRg = False
+            processChar = False
+            encoding = []
+            fontType = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Subtype"]
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1073]:
```suggestion
            fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
```

File: PyPDF2/_page.py [hunk 5]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+            if "/Encoding" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1074]:
```suggestion
            if "/Encoding" in pdf.pages[0]["/Resources"]["/Font"][font_name]:
```

File: PyPDF2/_page.py [hunk 6]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+                enc= pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Encoding"].getObject()
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1075]:
```suggestion
                enc= pdf.pages[0]["/Resources"]["/Font"][font_name]["/Encoding"].get_object()
```

File: PyPDF2/_page.py [hunk 7]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+                    for o in enc['/Differences']:
+                        if isinstance(o,int):
+                            x = o
+                        else:
+                            try:
+                                encoding[x] = adobe_glyphs[o]
+                            except Exception:
+                                encoding[x] = o
+                            x += 1
+            if "/ToUnicode" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1091]:
```suggestion
            if "/ToUnicode" in pdf.pages[0]["/Resources"]["/Font"][font_name]:
```

File: PyPDF2/_page.py [hunk 8]
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+                cm = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/ToUnicode"].getData().decode('utf-8')
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1092]:
```suggestion
                cm = pdf.pages[0]["/Resources"]["/Font"][font_name]["/ToUnicode"].getData().decode('utf-8')
```

**Comment by MartinThoma** [comment]
I fixed the merge conflicts due to the recent PEP8 renamings + I adjusted the PR to the new syntax. There are still mypy issues though.

**Comment by MartinThoma** [comment]
Running the benchmark, I get:

```
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1141, in extract_text
    ft,cmap,cmap2 = buildCharMap(self.pdf,operands[0])
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1073, in buildCharMap
    fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/generic.py", line 519, in __getitem__
    return dict.__getitem__(self, key).get_object()
KeyError: '/R167'
```

**Comment by pubpub-zz** [comment]
@MartinThoma 
> Running the benchmark, I get:
> 
> ```
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1141, in extract_text
>     ft,cmap,cmap2 = buildCharMap(self.pdf,operands[0])
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1073, in buildCharMap
>     fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/generic.py", line 519, in __getitem__
>     return dict.__getitem__(self, key).get_object()
> KeyError: '/R167'
> ```

Thanks for the test and the report. I've fixed it; also I've introduced a test using testbench dataset. using urllib.request to get them. It's increasing the test duration. I've set a way to bypass it.

I think that this pull now should be proposed for testing 

PS : I merged with the new API and 2 functions are missing to prevent compatibility loss : getPage and getObject

**Comment by MartinThoma** [comment]
>     def read_from_stream(stream: StreamType, pdf: Any, forcedEncoding: Union[None,str,List[str],dict[int,str]] = None) -> "ArrayObject":  # PdfReader
> E   TypeError: 'type' object is not subscriptable

You need to change `dict[int,str]` to `Dict[int,str]` (capital `D`). The syntax you've used was only introduced in Python 3.9 or 3.10, I think

**Comment by MartinThoma** [comment]
A general question: If I see such things, should I directly adjust it in your PR? (I know some people love it and others hate it)

**Comment by pubpub-zz** [comment]
Thanks, for the advice. I've tried to improve myself.
I successed to fix it 😊
During your review, I would like to have your opinion : the solution I've used to prevent mypy errors with cast makes the code not so easy to read for me: can you give me some advice ?

**Comment by MartinThoma** [comment]
Glancing at the first result of [the benchmark](https://github.com/py-pdf/benchmarks) the results look amazing :heart_eyes: 

image

It will take a while until the benchmark completed. I'll share the results later :-)

**Comment by MartinThoma** [comment]
@pubpub-zz Could you please have a look at https://arxiv.org/pdf/1601.03642.pdf ? The text extraction now completely fails on this one. It only extracts Asian characters, although the document doesn't have a single Asian character:

image

**Comment by MartinThoma** [comment]
@pubpub-zz I have the impression you're trying too much at one time here. For example, I can identify at least two different improvements in this PR:

* Spaces (spaceScale)
* CharMap support (buildCharMap)

I guess there are even more.

The spaces improvement seems to add already quite a bit of value. It also seems ready.

Although the CharMap support is the bigger topic, it's also one that is not quite ready yet (due to the Chinese-character issue mentioned above).

I suggest to do those step-by-step. I admire that you're having so many ideas and that you're able to implement them! However, for me as a maintainer it's easier to review smaller chunks. For you as a contributor it might also be more rewarding, because your changes ultimately get merged faster

**Comment by MartinThoma** [comment]
> the solution I've used to prevent mypy errors with cast makes the code not so easy to read for me

I'll have a look :-)

File: PyPDF2/_page.py [hunk 9]
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+        def buildCharMap(font_name : str)->Tuple[str,Dict,Dict]:
+            mapDict : Any = {}
+            processRg : bool = False
+            processChar : bool = False
+            encoding : List[str] = []
+            fontType :str = cast(NameObject, cast(DictionaryObject,
+                                 cast(DictionaryObject,
+                                      cast(DictionaryObject, self["/Resources"])["/Font"]
+                                 )[font_name]
+                            )["/Subtype"])
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1088]: Cast is telling mypy to ignore what it thinks the type is. Typically you only need one cast per line (if you need any at all)

```suggestion
            fontType :str = cast(NameObject, self["/Resources"]["/Font"][font_name]["/Subtype"])
```

File: PyPDF2/_page.py [hunk 10]
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+                        lst = [x for x in l.split(" ") if x]
+                        a = int(lst[0], 16)
+                        mapDict[a] = unhexlify(lst[1]).decode("utf-16-be")
+            return fontType, dict(zip(range(256), encoding)), "".maketrans(mapDict)
+        # ------- end of buildCharmap ------
+        text: str = ""
+        output: str = ""
+        cmaps: Dict[str, Tuple[str, Dict[int,str],Dict[str,str]]]={}
+        if "/Font" in cast(DictionaryObject,self["/Resources"]):
+            for f in cast(DictionaryObject,cast(DictionaryObject,self["/Resources"])["/Font"]):
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1183]:
```suggestion
        resources_dict = cast(DictionaryObject,self["/Resources"])
        if "/Font" in resources_dict:
            for f in cast(DictionaryObject, resources_dict["/Font"]):
```

File: PyPDF2/_page.py [hunk 11]
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
         if not isinstance(content, ContentStream):
-            content = ContentStream(content, self.pdf)
+            content = ContentStream(content, self.pdf, "charmap")
         # Note: we check all strings are TextStringObjects.  ByteStringObjects
         # are strings where the byte->string encoding was unknown, so adding
         # them to the text here would be gibberish.
+
+        # charSize = 1.0
+        # charScale = 0.0
+        spaceScale = 1.0
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1195]: Nitpic: I try to get [PEP8](https://peps.python.org/pep-0008/) compliant code. This makes it easier to read as it's one style.

I would still merge any PR that violates PEP8, but I would later clean those lines up.

File: PyPDF2/generic.py [hunk 12]
```diff
@@ -200,7 +200,7 @@ def writeToStream(
         self.write_to_stream(stream, encryption_key)
 
     @staticmethod
-    def read_from_stream(stream: StreamType, pdf: Any) -> "ArrayObject":  # PdfReader
+    def read_from_stream(stream: StreamType, pdf: Any, forcedEncoding: Union[None,str,List[str],Dict[int,str]] = None) -> "ArrayObject":  # PdfReader
```
Comment (MartinThoma) [inline PyPDF2/generic.py:203]: forcedEncoding seems never to be assigned. I guess that is one part that is still a TODO? What does this do?

File: PyPDF2/_page.py [hunk 13]
```diff
@@ -1069,48 +1075,185 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
...
+            processChar : bool = False
+            encoding : List[str] = []
+            fontType :str = cast(NameObject, self["/Resources"]["/Font"][font_name]["/Subtype"])
+            if "/Encoding" in cast(DictionaryObject, cast(DictionaryObject, cast(DictionaryObject, self["/Resources"])["/Font"])[font_name]):
+                enc = cast(DictionaryObject,
+                           cast(DictionaryObject,
+                                cast(DictionaryObject,
+                                     cast(DictionaryObject, self["/Resources"])["/Font"]
+                                )[font_name]
+                            )["/Encoding"]
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1091]:
```suggestion
            resources_dict = cast(DictionaryObject, self["/Resources"])
            fontType :str = cast(NameObject, resources_dict["/Font"][font_name]["/Subtype"])
            if "/Encoding" in resources_dict["/Font"][font_name]:
                enc = resources_dict["/Font"][font_name]["/Encoding"]
```

**Comment by MartinThoma** [comment]
Results of the benchmark run (the numbers are the documents):

1. 89% ➔ 89% -- big improvement!
2. 79% ➔ 85% -- big improvement!
3. 89% -> 71% :cry: -- quite a lot of spacing issues + greek characters (maybe cmap?)
4. 91% -> 93%
5. 95% -> 93% -- Meld fails to show a propper diff
6. 92% -> 93% -- Meld fails to show a propper diff
7. 90% -> 88% -- Meld fails to show a propper diff
8. 89% -> 91% -- Meld fails to show a propper diff
9. 97% -> 95% -- Meld fails to show a propper diff
10. 90% -> 90%
11. 97% -> 95%
12. 91% -> 92% -- Meld fails to show a propper diff
13. 0% -> 0%
14. 94% -> 94% -- Meld fails to show a propper diff

I came a big step closer: It seems as if my diff tool ( https://meldmerge.org/ ) had issues opening the file

**Comment by MartinThoma** [comment]
It seems like there are some extracted characters that are not in UTF-8:

image

Hence meld falls back to UTF-16 which is just wrong. I'm uncertain if PyPDF2 can (or should) do anything here.

**Comment by pubpub-zz** [comment]
> Regarding the Asian character issue: It seems like there are some extracted characters that are not in UTF-8:
> 
> image
> 
> Hence meld falls back to UTF-16 which is just wrong. I'm uncertain if PyPDF2 can (or should) do anything here.

@MartinThoma
This is my code:
`
import io,urllib.request,PyPDF2;p = PyPDF2.PdfReader(io.BytesIO(urllib.request.urlopen("https://arxiv.org/pdf/1601.03642.pdf").read()));e=p.pages[0].extract_text();print(e);f=open("e:/extracted.txt","wb");f.write(e.encode("utf-8"));f.close()
`
I'm getting some good results.

**Comment by pubpub-zz** [comment]
@MartinThoma ,
Unless I'm wrong your proposal seems to not make mypy happy 😕

**Comment by MartinThoma** [comment]
@pubpub-zz You're right, I missed some things: https://github.com/py-pdf/PyPDF2/pull/881#discussion_r884123778

**Comment by pubpub-zz** [comment]
@MartinThoma,
My latest status
a) about issues with asian characters : I've not been able to reproduce the issue.
my code in python to get the file:
`import io,urllib.request,PyPDF2;p = PyPDF2.PdfReader(io.BytesIO(urllib.request.urlopen("https://arxiv.org/pdf/2201.00151.pdf").read()));e=p.pages[0].extract_text();print(e);f=open('e:/out.txt','w',encoding='utf-8');f.write(e);f.close()`
the result of comparison between out.txt(left) and copy/paste from adobe reader DC: 
image

b) spacing:
I've sligthtly modified the code to allow to pass the space width criteria. I' still do not understand some case where a lower width is sometimes interpreted as a space whereas is some cases not.

c) mypy.
Your proposal seems to have reintroduced some error reporting. Can you check on your side and reajust enventually?

d)benchmark
Can you tell me how are you obtaining the percentage result?

**Comment by MartinThoma** [comment]
I've extracted the spacing improvement heuristic to https://github.com/py-pdf/PyPDF2/pull/922 . That should be possible to get merged really soon

**Comment by MartinThoma** [comment]
I've just fixed the merge conflict I've introduced

**Comment by MartinThoma** [comment]
@pubpub-zz Here is the simplification for the casts: https://github.com/py-pdf/PyPDF2/pull/924/commits/c4887349924127a46852e80397909c9dffa28c04 - essentially just assigning variable names to intermediate results

**Comment by LucasWolfgang** [comment]
Hey man, just want to say thank you for you hard work on this one. I got some interest on do this improved myself and end up finding this PR. Really nice work, @pubpub-zz. Are you checking list your past and future updates somewhere? I would like to follow and see if I can help to add html features to it (while avoiding disturb your work)

**Comment by pubpub-zz** [comment]
> 

@MartinThoma, your proposal is quite heavy to read. May I propose this solution?
`            font_type: str = self["/Resources"]["/Font"][font_name]["/Subtype"]    # type: ignore`

**Comment by MartinThoma** [comment]
That is completely fine for me as well :-)

**Comment by pubpub-zz** [comment]
> > 
> 
> @MartinThoma, your proposal is quite heavy to read. May I propose this solution? ` font_type: str = self["/Resources"]["/Font"][font_name]["/Subtype"] # type: ignore`

@LucasWolfgang, thanks for your nice comments. There will be still some improvement to be done. Don't know i will be able to release new changes.
What do you intend to propose (font? style? position?)

**Comment by LucasWolfgang** [comment]
@pubpub-zz , I was thinking of implemting the functionalities available in PyMuPDF's [get_text](https://pymupdf.readthedocs.io/en/latest/tutorial.html#extracting-text-and-images). Of course it would be a LOT of work. So I would start fist by just creating the blocks (paragraphs) and continue from there to extract font, style and Images (as Base64 or maybe as references) a PR at a time. About positions, I am still unsure how I would use it besides ordering the blocks.

**Comment by pubpub-zz** [comment]
maybe what you could do would be to branch from my dev branch https://github.com/pubpub-zz/PyPDF2/tree/extractText
If you had an as_html boolean parameter to add so if branches to be able keep text and html at the same time. This solution may help you to merge my changes later

**Comment by MartinThoma** [comment]
@pubpub-zz #924 got merged :tada: I'd make a minor release (2.1.0) on Monday.

**Comment by MartinThoma** [comment]
I've moved the buildCharMap function to its own module: https://github.com/py-pdf/PyPDF2/commit/4baedb2054648b9db70c2fbed82b6902aea52365 - I hope this simplifies future PRs / makes it easier for me to understand what they are actually doing.

**Pull request closed without merging**